{"lang":"Javascript","fullPath":"/Global_Objects/Error","category":"Global_Objects","title":"Error","url":"/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error","markdown":"# Error\n\n**`Error`** objects are thrown when runtime errors occur. The `Error` object can also be used as a base object for user-defined exceptions. See below for standard built-in error types.\n\n## [Description](#description)\n\nRuntime errors result in new `Error` objects being created and thrown.\n\n`Error` is a [serializable object](/en-US/docs/Glossary/Serializable%5Fobject), so it can be cloned with [structuredClone()](/en-US/docs/Web/API/structuredClone) or copied between [Workers](/en-US/docs/Web/API/Worker) using [postMessage()](/en-US/docs/Web/API/Worker/postMessage \"postMessage()\").\n\n### [Error types](#error%5Ftypes)\n\nBesides the generic `Error` constructor, there are other core error constructors in JavaScript. For client-side exceptions, see [Exception handling statements](/en-US/docs/Web/JavaScript/Guide/Control%5Fflow%5Fand%5Ferror%5Fhandling#exception%5Fhandling%5Fstatements).\n\n[EvalError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/EvalError)\n\nCreates an instance representing an error that occurs regarding the global function [eval()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/eval).\n\n[RangeError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/RangeError)\n\nCreates an instance representing an error that occurs when a numeric variable or parameter is outside its valid range.\n\n[ReferenceError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/ReferenceError)\n\nCreates an instance representing an error that occurs when de-referencing an invalid reference.\n\n[SyntaxError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/SyntaxError)\n\nCreates an instance representing a syntax error.\n\n[TypeError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/TypeError)\n\nCreates an instance representing an error that occurs when a variable or parameter is not of a valid type.\n\n[URIError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/URIError)\n\nCreates an instance representing an error that occurs when [encodeURI()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/encodeURI) or [decodeURI()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/decodeURI) are passed invalid parameters.\n\n[AggregateError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/AggregateError)\n\nCreates an instance representing several errors wrapped in a single error when multiple errors need to be reported by an operation, for example by [Promise.any()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Promise/any).\n\n[InternalError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/InternalError) Non-standard \n\nCreates an instance representing an error that occurs when an internal error in the JavaScript engine is thrown. E.g. \"too much recursion\".\n\n## [Constructor](#constructor)\n\n[Error()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/Error)\n\nCreates a new `Error` object.\n\n## [Static methods](#static%5Fmethods)\n\n`Error.captureStackTrace()` Non-standard \n\nA non-standard V8 function that creates the [stack](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/stack) property on an Error instance.\n\n`Error.stackTraceLimit` Non-standard \n\nA non-standard V8 numerical property that limits how many stack frames to include in an error stacktrace.\n\n`Error.prepareStackTrace()` Non-standard Optional\n\nA non-standard V8 function that, if provided by usercode, is called by the V8 JavaScript engine for thrown exceptions, allowing the user to provide custom formatting for stacktraces.\n\n## [Instance properties](#instance%5Fproperties)\n\nThese properties are defined on `Error.prototype` and shared by all `Error` instances.\n\n[Error.prototype.constructor](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Object/constructor)\n\nThe constructor function that created the instance object. For `Error` instances, the initial value is the [Error](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/Error) constructor.\n\n[Error.prototype.name](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/name)\n\nRepresents the name for the type of error. For `Error.prototype.name`, the initial value is `\"Error\"`. Subclasses like [TypeError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/TypeError) and [SyntaxError](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/SyntaxError) provide their own `name` properties.\n\n[Error.prototype.stack](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/stack) Non-standard \n\nA non-standard property for a stack trace.\n\nThese properties are own properties of each `Error` instance.\n\n[cause](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/cause)\n\nError cause indicating the reason why the current error is thrown — usually another caught error. For user-created `Error` objects, this is the value provided as the `cause` property of the constructor's second argument.\n\n[columnNumber](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/columnNumber) Non-standard \n\nA non-standard Mozilla property for the column number in the line that raised this error.\n\n[fileName](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/fileName) Non-standard \n\nA non-standard Mozilla property for the path to the file that raised this error.\n\n[lineNumber](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/lineNumber) Non-standard \n\nA non-standard Mozilla property for the line number in the file that raised this error.\n\n[message](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/message)\n\nError message. For user-created `Error` objects, this is the string provided as the constructor's first argument.\n\n## [Instance methods](#instance%5Fmethods)\n\n[Error.prototype.toString()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/toString)\n\nReturns a string representing the specified object. Overrides the [Object.prototype.toString()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Object/toString) method.\n\n## [Examples](#examples)\n\n### [Throwing a generic error](#throwing%5Fa%5Fgeneric%5Ferror)\n\n Usually you create an `Error` object with the intention of raising it using the [throw](/en-US/docs/Web/JavaScript/Reference/Statements/throw) keyword. You can handle the error using the [try...catch](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch) construct:\n\n```\ntry {\n  throw new Error(\"Whoops!\");\n} catch (e) {\n  console.error(`${e.name}: ${e.message}`);\n}\n\n```\n\n### [Handling a specific error type](#handling%5Fa%5Fspecific%5Ferror%5Ftype)\n\nYou can choose to handle only specific error types by testing the error type with the error's [constructor](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Object/constructor) property or, if you're writing for modern JavaScript engines, [instanceof](/en-US/docs/Web/JavaScript/Reference/Operators/instanceof) keyword:\n\n```\ntry {\n  foo.bar();\n} catch (e) {\n  if (e instanceof EvalError) {\n    console.error(`${e.name}: ${e.message}`);\n  } else if (e instanceof RangeError) {\n    console.error(`${e.name}: ${e.message}`);\n  }\n  // etc.\n  else {\n    // If none of our cases matched leave the Error unhandled\n    throw e;\n  }\n}\n\n```\n\n### [Differentiate between similar errors](#differentiate%5Fbetween%5Fsimilar%5Ferrors)\n\nSometimes a block of code can fail for reasons that require different handling, but which throw very similar errors (i.e. with the same type and message).\n\n If you don't have control over the original errors that are thrown, one option is to catch them and throw new `Error` objects that have more specific messages. The original error should be passed to the new `Error` in the constructor's [options](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/Error#options) parameter as its `cause` property. This ensures that the original error and stack trace are available to higher-level try/catch blocks.\n\nThe example below shows this for two methods that would otherwise fail with similar errors (`doFailSomeWay()` and `doFailAnotherWay()`):\n\n```\nfunction doWork() {\n  try {\n    doFailSomeWay();\n  } catch (err) {\n    throw new Error(\"Failed in some way\", { cause: err });\n  }\n  try {\n    doFailAnotherWay();\n  } catch (err) {\n    throw new Error(\"Failed in another way\", { cause: err });\n  }\n}\n\ntry {\n  doWork();\n} catch (err) {\n  switch (err.message) {\n    case \"Failed in some way\":\n      handleFailSomeWay(err.cause);\n      break;\n    case \"Failed in another way\":\n      handleFailAnotherWay(err.cause);\n      break;\n  }\n}\n\n```\n\n**Note:** If you are making a library, you should prefer to use error cause to discriminate between different errors emitted — rather than asking your consumers to parse the error message. See the [error cause page](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/cause#providing%5Fstructured%5Fdata%5Fas%5Fthe%5Ferror%5Fcause) for an example.\n\n[Custom error types](#custom%5Ferror%5Ftypes) can also use the [cause](#error.prototype.cause) property, provided the subclasses' constructor passes the `options` parameter when calling `super()`. The `Error()` base class constructor will read `options.cause` and define the `cause` property on the new error instance.\n\n```\nclass MyError extends Error {\n  constructor(message, options) {\n    // Need to pass `options` as the second parameter to install the \"cause\" property.\n    super(message, options);\n  }\n}\n\nconsole.log(new MyError(\"test\", { cause: new Error(\"cause\") }).cause);\n// Error: cause\n\n```\n\n### [Custom error types](#custom%5Ferror%5Ftypes)\n\nYou might want to define your own error types deriving from `Error` to be able to `throw new MyError()` and use `instanceof MyError` to check the kind of error in the exception handler. This results in cleaner and more consistent error handling code.\n\nSee [\"What's a good way to extend Error in JavaScript?\"](https://stackoverflow.com/questions/1382107/whats-a-good-way-to-extend-error-in-javascript) on StackOverflow for an in-depth discussion.\n\n**Warning:** Builtin subclassing cannot be reliably transpiled to pre-ES6 code, because there's no way to construct the base class with a particular `new.target` without [Reflect.construct()](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Reflect/construct). You need [additional configuration](https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend) or manually call [Object.setPrototypeOf(this, CustomError.prototype)](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Object/setPrototypeOf) at the end of the constructor; otherwise, the constructed instance will not be a `CustomError` instance. See [the TypeScript FAQ](https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work) for more information.\n\n**Note:** Some browsers include the `CustomError` constructor in the stack trace when using ES2015 classes.\n\n```\nclass CustomError extends Error {\n  constructor(foo = \"bar\", ...params) {\n    // Pass remaining arguments (including vendor specific ones) to parent constructor\n    super(...params);\n\n    // Maintains proper stack trace for where our error was thrown (only available on V8)\n    if (Error.captureStackTrace) {\n      Error.captureStackTrace(this, CustomError);\n    }\n\n    this.name = \"CustomError\";\n    // Custom debugging information\n    this.foo = foo;\n    this.date = new Date();\n  }\n}\n\ntry {\n  throw new CustomError(\"baz\", \"bazMessage\");\n} catch (e) {\n  console.error(e.name); // CustomError\n  console.error(e.foo); // baz\n  console.error(e.message); // bazMessage\n  console.error(e.stack); // stacktrace\n}\n\n```\n\n## [Specifications](#specifications)\n\n| Specification                                                                                                                           |\n| --------------------------------------------------------------------------------------------------------------------------------------- |\n| [ECMAScript Language Specification  \\# sec-error-objects](https://tc39.es/ecma262/multipage/fundamental-objects.html#sec-error-objects) |\n\n## [Browser compatibility](#browser%5Fcompatibility)\n\nBCD tables only load in the browser\n\n## [See also](#see%5Falso)\n\n* [A polyfill of Error](https://github.com/zloirock/core-js#ecmascript-error) with modern behavior like support `cause` is available in [core-js](https://github.com/zloirock/core-js)\n* [throw](/en-US/docs/Web/JavaScript/Reference/Statements/throw)\n* [try...catch](/en-US/docs/Web/JavaScript/Reference/Statements/try...catch)\n* The [V8 documentation](https://v8.dev/docs/stack-trace-api) for `Error.captureStackTrace()`, `Error.stackTraceLimit`, and `Error.prepareStackTrace()`.\n\n### Found a content problem with this page?\n\n* [Edit the page on GitHub](https://github.com/mdn/content/edit/main/files/en-us/web/javascript/reference/global%5Fobjects/error/index.md \"This will take you to GitHub, where you'll need to sign in first.\").\n* [Report the content issue](https://github.com/mdn/content/issues/new?template=page-report.yml&mdn-url=https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FJavaScript%2FReference%2FGlobal%5FObjects%2FError&metadata=%3C%21--+Do+not+make+changes+below+this+line+--%3E%0A%3Cdetails%3E%0A%3Csummary%3EPage+report+details%3C%2Fsummary%3E%0A%0A%2A+Folder%3A+%60en-us%2Fweb%2Fjavascript%2Freference%2Fglobal%5Fobjects%2Ferror%60%0A%2A+MDN+URL%3A+https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FJavaScript%2FReference%2FGlobal%5FObjects%2FError%0A%2A+GitHub+URL%3A+https%3A%2F%2Fgithub.com%2Fmdn%2Fcontent%2Fblob%2Fmain%2Ffiles%2Fen-us%2Fweb%2Fjavascript%2Freference%2Fglobal%5Fobjects%2Ferror%2Findex.md%0A%2A+Last+commit%3A+https%3A%2F%2Fgithub.com%2Fmdn%2Fcontent%2Fcommit%2F01bf58c144237c1fc75cf7135491346778591f09%0A%2A+Document+last+modified%3A+2023-04-08T17%3A33%3A52.000Z%0A%0A%3C%2Fdetails%3E \"This will take you to GitHub to file a new issue.\").\n* [View the source on GitHub](https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global%5Fobjects/error/index.md?plain=1 \"Folder: en-us/web/javascript/reference/global_objects/error (Opens in a new tab)\").\nWant to get more involved? [Learn how to contribute](https://github.com/mdn/content/blob/main/CONTRIBUTING.md \"This will take you to our contribution guidelines on GitHub.\").\n\nThis page was last modified on Apr 8, 2023 by [MDN contributors](/en-US/docs/Web/JavaScript/Reference/Global%5FObjects/Error/contributors.txt)."}